homeassistant/components/icloud.py

```diff
@@ -0,0 +1,956 @@
+        ICLOUDTRACKERS[account] = icloudaccount
+        if ICLOUDTRACKERS[account].api is not None:
+            for device in ICLOUDTRACKERS[account].devices:
+                iclouddevice = ICLOUDTRACKERS[account].devices[device]
+                devicename = iclouddevice.devicename.lower()
+                track_state_change(hass,
+                                   'device_tracker.' + devicename,
+                                   iclouddevice.devicechanged)
+
+    if not ICLOUDTRACKERS:
+        _LOGGER.error("No ICLOUDTRACKERS added")
+        return False
+
+    randomseconds = random.randint(10, 59)
+
+    def lost_iphone(call):
+        """Call the lost iphone function if the device is found."""
+        accountname = call.data.get(ATTR_ACCOUNTNAME)
+        devicename = call.data.get(ATTR_DEVICENAME)
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].lost_iphone(devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].lost_iphone(devicename)
+
+    hass.services.register(DOMAIN, 'lost_iphone',
+                           lost_iphone)
+
+    def update_icloud(call):
+        """Call the update function of an icloud account."""
+        accountname = call.data.get(ATTR_ACCOUNTNAME)
+        devicename = call.data.get(ATTR_DEVICENAME)
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].update_icloud(devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].update_icloud(devicename)
+    hass.services.register(DOMAIN,
+                           'update_icloud', update_icloud)
+
+    def keep_alive(now):
+        """Keep the api logged in of all account."""
+        for accountname in ICLOUDTRACKERS:
+            try:
+                ICLOUDTRACKERS[accountname].keep_alive()
+            except ValueError:
+                _LOGGER.error("something went wrong for account %s, " +
+                              "retrying in a minute", accountname)
+
+    track_utc_time_change(
+        hass, keep_alive,
+        second=randomseconds
+    )
+
+    def setinterval(call):
+        """Call the update function of an icloud account."""
+        accountname = call.data.get(ATTR_ACCOUNTNAME)
+        interval = call.data.get(ATTR_INTERVAL)
+        devicename = call.data.get(ATTR_DEVICENAME)
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].setinterval(interval, devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].setinterval(interval, devicename)
+
+    hass.services.register(DOMAIN,
+                           'setinterval', setinterval)
+
+    # Tells the bootstrapper that the component was successfully initialized
+    return True
+
+
+class IDevice(Entity):  # pylint: disable=too-many-instance-attributes
```

Comment:
There shouldn't be ever 1 component with 3 different entities. Spread it out. Keep the device tracker platform a device tracker platform but have it depend on this code, hook it in .